Fix when _parallel option is a variable, workflow fails

szyn opened the pull request:
Follow up of #1413 

#1413 breaks backward compatibility. There are issues as follows:

1.  When using a variable for _parallel in a subtask configuration, the task fails because it cannot be evaluated. 
    - Note that we can't use a variable for _parallel in group tasks, but it can be specified in subtasks.
1.  As for loop operators (loop, for_each, td_for_each, for_range), OperatorManager can't correctly catch errors if it only checks for the existence of the _parallel parameter in their operator factory.

I want to fix those as follows:

1. Update the value of _parallel is correctly evaluated, even if a variable is specified. But it can only use in subtasks.
1. Move ParallelControl class to `digdag-plugin-utils` project, and in the loop's OperatorFactory, ParallelControl will check for errors when the format is wrong.

And I also added the missing test cases.

komamitsu commented:
> There are issues as follows:

@szyn Can you describe how to reproduce the issues?

szyn commented:
Yes, sure. First of all, I used digdag built with [c57bf873054f644b36365b04d078b77d8567a17f](https://github.com/treasure-data/digdag/tree/c57bf873054f644b36365b04d078b77d8567a17f)


The following steps can reproduce this.

```bash
$ git rev-parse HEAD
c57bf873054f644b36365b04d078b77d8567a17f

$ ./gradlew cli
Publication mavenJava not found in project :.
Publication mavenJava not found in project :digdag-docs.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.3/userguide/command_line_interface.html#sec:command_line_warnings
```


```bash
### Prepare the project
$ cat ~/work/issue-1449/parallel_variable_loop.dig
_export:
  parallel: true

+loop:
  loop>: 2
  _parallel: ${parallel}
  _do:
    +task1:
      echo>: loop${i},task1
    +task2:
      echo>: loop${i},task2
    +task3:
      echo>: loop${i},task3

$ cd /path/to/digdag

### Start digdag-server
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar server -m --task-log /tmp/digdag-task-logs
2020-08-31 17:50:37 +0900: Digdag v0.9.38
2020-08-31 17:50:38 +0900 [INFO] (main): secret encryption engine: disabled
2020-08-31 17:50:38 +0900 [INFO] (main): XNIO version 3.3.6.Final
2020-08-31 17:50:38 +0900 [INFO] (main): XNIO NIO Implementation Version 3.3.6.Final
2020-08-31 17:50:38 +0900 [INFO] (main): Starting server on 127.0.0.1:65432
2020-08-31 17:50:38 +0900 [INFO] (main): Bound on 127.0.0.1:65432 (api)
..snip..

### Push the project
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar push issue-1449 --project ~/work/issue-1449
2020-08-31 17:50:55 +0900: Digdag v0.9.38
Creating .digdag/tmp/archive-1456876824283518605.tar.gz...
  Archiving parallel_variable_loop.dig
Workflows:
  parallel_variable_loop.dig
Uploaded:
  id: 1
  name: issue-1449
  revision: 916f4df2-d750-455a-88c7-a51e142fc766
  archive type: db
  project created at: 2020-08-31T08:50:57Z
  revision updated at: 2020-08-31T08:50:57Z

Use `digdag workflows` to show all workflows.

### Start the workflow (parallel_variable_loop)
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar start issue-1449 parallel_variable_loop --session now
2020-08-31 17:51:29 +0900: Digdag v0.9.38
Started a session attempt:
  session id: 1
  attempt id: 1
  uuid: ba0b5bb9-8040-4eba-b257-b3f4c715a653
  project: issue-1449
  workflow: parallel_variable_loop
  session time: 2020-08-31 08:51:30 +0000
  retry attempt name:
  params: {}
  created at: 2020-08-31 17:51:30 +0900

* Use `digdag session 1` to show session status.
* Use `digdag task 1` and `digdag log 1` to show task status and logs.

$ ./pkg/digdag-0.10.0-SNAPSHOT.jar task 1
2020-08-31 17:51:45 +0900: Digdag v0.9.38
   id: 1
   name: +parallel_variable_loop
   state: group_error
   started:
   updated: 2020-08-31 17:51:31 +0900
   config: {"_export":{"parallel":true}}
   parent: null
   upstreams: []
   export params: {"parallel":true}
   store params: {}
   state params: {}

   id: 2
   name: +parallel_variable_loop+loop
   state: error
   started: 2020-08-31 17:51:30 +0900
   updated: 2020-08-31 17:51:31 +0900
   config: {"loop>":2,"_parallel":"${parallel}","_do":{"+task1":{"echo>":"loop${i},task1"},"+task2":{"echo>":"loop${i},task2"},"+task3":{"echo>":"loop${i},task3"}}}
   parent: 1
   upstreams: []
   export params: {}
   store params: {}
   state params: {}

   id: 3
   name: +parallel_variable_loop^failure-alert
   state: success
   started: 2020-08-31 17:51:31 +0900
   updated: 2020-08-31 17:51:31 +0900
   config: {"_type":"notify","_command":"Workflow session attempt failed"}
   parent: 1
   upstreams: []
   export params: {}
   store params: {}
   state params: {}

3 entries.

$ ./pkg/digdag-0.10.0-SNAPSHOT.jar log 1
./pkg/digdag-0.10.0-SNAPSHOT.jar log 1
2020-08-31 17:51:58 +0900: Digdag v0.9.38
2020-08-31 17:51:31.387 +0900 [INFO] (0041@[0:issue-1449]+parallel_variable_loop+loop) io.digdag.core.agent.OperatorManager: loop>: 2
2020-08-31 17:51:31.422 +0900 [INFO] (0041@[0:issue-1449]+parallel_variable_loop^failure-alert) io.digdag.core.agent.OperatorManager: type: notify
```

Note that this attempt will fail, but shows no errors because of #1301.

In addition to the above, this cannot reproduce with `run` command.

```bash
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar run -a parallel_variable_loop.dig --project ~/work/issue-1449
2020-08-31 17:55:26 +0900: Digdag v0.9.38
2020-08-31 17:55:26 +0900 [INFO] (main): Setting project path to /Users/suzuki.s/work/issue-1449
2020-08-31 17:55:26 +0900 [WARN] (main): Using a new session time 2020-08-31T00:00:00+00:00.
2020-08-31 17:55:26 +0900 [INFO] (main): Using session /Users/suzuki.s/work/issue-1449/.digdag/status/20200831T000000+0000.
2020-08-31 17:55:26 +0900 [INFO] (main): Starting a new session project id=1 workflow name=parallel_variable_loop session_time=2020-08-31T00:00:00+00:00
2020-08-31 17:55:27 +0900 [INFO] (0018@[0:default]+parallel_variable_loop+loop): loop>: 2
2020-08-31 17:55:27 +0900 [INFO] (0018@[0:default]+parallel_variable_loop^failure-alert): type: notify
error:
  * +parallel_variable_loop+loop:
    Invalid _parallel format: "true" (config)

Task state is saved at /Users/suzuki.s/work/issue-1449/.digdag/status/20200831T000000+0000 directory.
  * Use --session <daily | hourly | "yyyy-MM-dd[ HH:mm:ss]"> to not reuse the last session time.
  * Use --rerun, --start +NAME, or --goal +NAME argument to rerun skipped tasks.
```

szyn commented:
Yes, sure. First of all, I used digdag built with [c57bf873054f644b36365b04d078b77d8567a17f](https://github.com/treasure-data/digdag/tree/c57bf873054f644b36365b04d078b77d8567a17f)

```console
$ git rev-parse HEAD
c57bf873054f644b36365b04d078b77d8567a17f

$ ./gradlew cli
Publication mavenJava not found in project :.
Publication mavenJava not found in project :digdag-docs.

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.3/userguide/command_line_interface.html#sec:command_line_warnings
```

The following steps can reproduce this. This is only reproduced with server mode.

```console
### Prepare the project
$ cat ~/work/issue-1449/parallel_variable_loop.dig
_export:
  parallel: true

+loop:
  loop>: 2
  _parallel: ${parallel}
  _do:
    +task1:
      echo>: loop${i},task1
    +task2:
      echo>: loop${i},task2
    +task3:
      echo>: loop${i},task3

$ cd /path/to/digdag

### Start digdag-server
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar server -m --task-log /tmp/digdag-task-logs
2020-08-31 17:50:37 +0900: Digdag v0.9.38
2020-08-31 17:50:38 +0900 [INFO] (main): secret encryption engine: disabled
2020-08-31 17:50:38 +0900 [INFO] (main): XNIO version 3.3.6.Final
2020-08-31 17:50:38 +0900 [INFO] (main): XNIO NIO Implementation Version 3.3.6.Final
2020-08-31 17:50:38 +0900 [INFO] (main): Starting server on 127.0.0.1:65432
2020-08-31 17:50:38 +0900 [INFO] (main): Bound on 127.0.0.1:65432 (api)
..snip..

### Push the project
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar push issue-1449 --project ~/work/issue-1449
2020-08-31 17:50:55 +0900: Digdag v0.9.38
Creating .digdag/tmp/archive-1456876824283518605.tar.gz...
  Archiving parallel_variable_loop.dig
Workflows:
  parallel_variable_loop.dig
Uploaded:
  id: 1
  name: issue-1449
  revision: 916f4df2-d750-455a-88c7-a51e142fc766
  archive type: db
  project created at: 2020-08-31T08:50:57Z
  revision updated at: 2020-08-31T08:50:57Z

Use `digdag workflows` to show all workflows.

### Start the workflow (parallel_variable_loop)
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar start issue-1449 parallel_variable_loop --session now
2020-08-31 17:51:29 +0900: Digdag v0.9.38
Started a session attempt:
  session id: 1
  attempt id: 1
  uuid: ba0b5bb9-8040-4eba-b257-b3f4c715a653
  project: issue-1449
  workflow: parallel_variable_loop
  session time: 2020-08-31 08:51:30 +0000
  retry attempt name:
  params: {}
  created at: 2020-08-31 17:51:30 +0900

* Use `digdag session 1` to show session status.
* Use `digdag task 1` and `digdag log 1` to show task status and logs.

$ ./pkg/digdag-0.10.0-SNAPSHOT.jar task 1
2020-08-31 17:51:45 +0900: Digdag v0.9.38
   id: 1
   name: +parallel_variable_loop
   state: group_error
   started:
   updated: 2020-08-31 17:51:31 +0900
   config: {"_export":{"parallel":true}}
   parent: null
   upstreams: []
   export params: {"parallel":true}
   store params: {}
   state params: {}

   id: 2
   name: +parallel_variable_loop+loop
   state: error
   started: 2020-08-31 17:51:30 +0900
   updated: 2020-08-31 17:51:31 +0900
   config: {"loop>":2,"_parallel":"${parallel}","_do":{"+task1":{"echo>":"loop${i},task1"},"+task2":{"echo>":"loop${i},task2"},"+task3":{"echo>":"loop${i},task3"}}}
   parent: 1
   upstreams: []
   export params: {}
   store params: {}
   state params: {}

   id: 3
   name: +parallel_variable_loop^failure-alert
   state: success
   started: 2020-08-31 17:51:31 +0900
   updated: 2020-08-31 17:51:31 +0900
   config: {"_type":"notify","_command":"Workflow session attempt failed"}
   parent: 1
   upstreams: []
   export params: {}
   store params: {}
   state params: {}

3 entries.

$ ./pkg/digdag-0.10.0-SNAPSHOT.jar log 1
./pkg/digdag-0.10.0-SNAPSHOT.jar log 1
2020-08-31 17:51:58 +0900: Digdag v0.9.38
2020-08-31 17:51:31.387 +0900 [INFO] (0041@[0:issue-1449]+parallel_variable_loop+loop) io.digdag.core.agent.OperatorManager: loop>: 2
2020-08-31 17:51:31.422 +0900 [INFO] (0041@[0:issue-1449]+parallel_variable_loop^failure-alert) io.digdag.core.agent.OperatorManager: type: notify
```

Note that this attempt will fail, but shows no errors because of #1301.

In addition to the above, this cannot reproduce with `run` command.

```console
$ ./pkg/digdag-0.10.0-SNAPSHOT.jar run -a parallel_variable_loop.dig --project ~/work/issue-1449
2020-08-31 17:55:26 +0900: Digdag v0.9.38
2020-08-31 17:55:26 +0900 [INFO] (main): Setting project path to /path/to/work/issue-1449
2020-08-31 17:55:26 +0900 [WARN] (main): Using a new session time 2020-08-31T00:00:00+00:00.
2020-08-31 17:55:26 +0900 [INFO] (main): Using session /path/to/work/issue-1449/.digdag/status/20200831T000000+0000.
2020-08-31 17:55:26 +0900 [INFO] (main): Starting a new session project id=1 workflow name=parallel_variable_loop session_time=2020-08-31T00:00:00+00:00
2020-08-31 17:55:27 +0900 [INFO] (0018@[0:default]+parallel_variable_loop+loop): loop>: 2
2020-08-31 17:55:27 +0900 [INFO] (0018@[0:default]+parallel_variable_loop^failure-alert): type: notify
error:
  * +parallel_variable_loop+loop:
    Invalid _parallel format: "true" (config)

Task state is saved at /path/to/work/issue-1449/.digdag/status/20200831T000000+0000 directory.
  * Use --session <daily | hourly | "yyyy-MM-dd[ HH:mm:ss]"> to not reuse the last session time.
  * Use --rerun, --start +NAME, or --goal +NAME argument to rerun skipped tasks.
```

komamitsu commented on digdag-core/src/test/java/io/digdag/core/workflow/LoopOperatorFactory.java:
```diff
@@ -67,7 +69,7 @@ public TaskResult run()
 
-            if (params.has("_parallel")) {
+            if (ParallelControl.of(params).isParallel()) {
                 generated.set("_parallel", params.get("_parallel", JsonNode.class));
```
We're abstracting accesses to `_parallel` property by moving related implementations to  `ParallelControl`. Why don't you move this direct access to `_parallel` property to the class by adding like `ParallelControl#copyIfNeeded(Config src, Config dst)`?

szyn replied:
Thank you for pointing out, that's a really good point! Updated: `8b2915e`, `724c1a6`.

komamitsu commented on digdag-plugin-utils/src/main/java/io/digdag/util/ParallelControl.java:
```diff
@@ -6,16 +6,21 @@
 
 public class ParallelControl
 {
+    private static final String PARALLEL_KEY = "_parallel";
```
👍
